crates/sui-transactional-test-runner/src/test_adapter.rs

```diff
@@ -624,11 +632,163 @@ impl<'a> MoveTestAdapter<'a> for SuiTestAdapter<'a> {
+                            .collect();
+                        (modules, warnings_opt)
+                    }
+                    SyntaxChoice::IR => {
+                        let module = compile_ir_module(state, data.path())?;
+                        (vec![(None, module)], None)
+                    }
+                };
+                let output = self.upgrade_package(
+                    package_id,
+                    &modules,
+                    upgrade_capability,
+                    dependencies,
+                    sender,
+                    gas_budget,
+                )?;
+                match syntax {
+                    SyntaxChoice::Source => {
+                        let path = data.path().to_str().unwrap().to_owned();
+                        self.compiled_state()
+                            .add_with_source_file(modules, (path, data))
+                    }
+                    SyntaxChoice::IR => {
+                        let module = modules.pop().unwrap().1;
+                        self.compiled_state()
+                            .add_and_generate_interface_file(module);
+                    }
+                };
+                Ok(merge_output(warnings_opt, output))
+            }
+        }
+    }
+}
+
+fn merge_output(left: Option<String>, right: Option<String>) -> Option<String> {
+    match (left, right) {
+        (None, right) => right,
+        (left, None) => left,
+        (Some(mut left), Some(right)) => {
+            left.push_str(&right);
+            Some(left)
         }
     }
 }
 
 impl<'a> SuiTestAdapter<'a> {
+    fn upgrade_package(
+        &mut self,
+        package_id: ObjectID,
+        modules: &[(Option<Symbol>, CompiledModule)],
+        upgrade_capability: FakeID,
+        dependencies: Vec<String>,
+        sender: String,
+        gas_budget: Option<u64>,
+    ) -> anyhow::Result<Option<String>> {
+        let modules = modules
+            .iter()
+            .map(|(_, module)| {
+                let mut module_bytes = vec![];
+                module.serialize(&mut module_bytes)?;
+                Ok(module_bytes)
+            })
+            .collect::<anyhow::Result<Vec<Vec<u8>>>>()?;
+        let gas_budget = gas_budget.unwrap_or(DEFAULT_GAS_BUDGET);
+
+        let mut dependencies: Vec<_> = dependencies
+            .into_iter()
+            .map(|d| {
+                let Some(addr) = self.compiled_state.named_address_mapping.get(&d) else {
+                    bail!("There is no published module address corresponding to name address {d}");
+                };
+                let id: ObjectID = addr.into_inner().into();
+                Ok(id)
+            })
+            .collect::<Result<_, _>>()?;
+        dependencies.extend(BuiltInFramework::all_package_ids());
+
+        let mut builder = ProgrammableTransactionBuilder::new();
+
+        SuiValue::Object(upgrade_capability).into_argument(&mut builder, self)?; // Argument::Input(0)
```

Comment:
Is there not a way in the builder to pass in this input and get `Argument::Input(0)` out as a return value?  That would be preferable to using comments for bookkeeping.

Reply:
I couldn't find an obvious way, and borrowed this from some other code (don't remember where I saw it). What complicates getting the return value is that `upgrade_capability` here is a `FakeID`, otherwise I would do something like `let cap = builder.obj(ObjectArg::ImmOrOwnedObject(upgrade_capability))`.

(the fact that this is a side-effecting call also bothers me a lot, which is why I added the comment).

Reply:
Yeah it's a bit strange with the whole fake ID thing. Really we would need something that converts it into a `CallArg` and then you could call `builder.call_arg` or `input` or whatever it's called.